src/signature.c

```diff
@@ -435,6 +435,14 @@ static CK_RV p11prov_sig_pss_restrictions(P11PROV_SIG_CTX *sigctx,
         int num_mechs = allowed_mechs->ulValueLen;
         bool allowed = false;
 
+        if (num_mechs == 0) {
+            /* It makes no sense to return 0 allowed mechanisms for a key,
+             * this just means the token is bogus, let's ignore the check
+             * and try the operation and see what happens */
```

Comment:
I would say that the `CKA_ALLOWED_MECHANISMS` is underspecified. I think when I implemented this in SoftHSM, returning empty array was closest thing to expressing "anything is allowed"/"there are no limitations".

Reply:
the specification says:
```
A list of mechanisms allowed to be used with this key. The number of mechanisms in the array is the ulValueLen component of the attribute divided by the size of CK_MECHANISM_TYPE.
```

If you  return an attribute at all you are indicating this list, if you return an empty list you are saying nothing is allowed ... I think this is a bug in softhsm.

Reply:
But I am glade for this bug, as I just saw that my code to then go through the list is broken :-D

Reply:
Just to be clear, the correct behavior to mean "anythign is allowed" is to not return this attribute at all.
Alternatively return all the known mechanisms related to this key.

Reply:
I think it is still not completely clear from the specs. Right, there is s paragraph:

> 1.     If the specified attribute (i.e., the attribute specified by the type field) for the object cannot be revealed because the object is sensitive or unextractable, then the ulValueLen field in that triple is modified to hold the value CK_UNAVAILABLE_INFORMATION.

but it talks only about "technical" issues to read the attribute. But thinking about it further, you are probably right that getting `CK_UNAVAILABLE_INFORMATION` would be better.

Reply:
oh, The allowed mechanisms was not the part I was implementing, I think. Its just the list provided by the one who generated the key or imported the key through pkcs11 so if there were no provided in the templates, it returns just the empty array ... so in the end, its probably the creator responsiblity, but yeah, softhsm should probably handle this corner case better.

Reply:
Most key generation code does not want to restrict algorithms so they set nothing, and in that case softhsm should return CK_UNAVAILABLE_INFORMATION.
I think returning 0 length array is never expected and improper behavior.